Change scheme for the google maps api call to https and include an api key if one was set

matrikular opened the pull request:
#### Summary of Changes

This PR changes the api scheme / protocol for the geocodeAddress call to https instead of http. Currently the JGoogleEmbedMaps::geocodeAddress() api call uses http.

Though, e.g. if you call the api url directly, google redirects you to https, an exception will be returned from google when using the api together with an api key without using https.

> object(JHttpResponse)[1247]
>   public 'code' => int 200
>   public 'headers' =>
>     array (size=12)
>       'Content-Type' => string 'application/json; charset=UTF-8' (length=31)
>       'Date' => string 'Thu, 12 May 2016 10:18:22 GMT' (length=29)
>       'Pragma' => string 'no-cache' (length=8)
>       'Expires' => string 'Fri, 01 Jan 1990 00:00:00 GMT' (length=29)
>       'Cache-Control' => string 'no-cache, must-revalidate' (length=25)
>       'Access-Control-Allow-Origin' => string '*' (length=1)
>       'Server' => string 'mafe' (length=4)
>       'X-XSS-Protection' => string '1; mode=block' (length=13)
>       'X-Frame-Options' => string 'SAMEORIGIN' (length=10)
>       'Accept-Ranges' => string 'none' (length=4)
>       'Vary' => string 'Accept-Language,Accept-Encoding' (length=31)
>       'Transfer-Encoding' => string 'chunked' (length=7)
>   public 'body' => string '{
>    "error_message" : "Requests to this API must be over SSL. Load the API with \"https://\" instead of \"http://\".",
>    "results" : [],
>    "status" : "REQUEST_DENIED"
> }
> ' (length=172)

Speaking of the api key; The google library allows you to set an api key that will be used in the javascript rendering part but not in the geocodeAddress() method. This PR also adds a key to the request if present so that you won't run into your quota limit any time soon.
#### Testing Instructions

Download and install the test modul I attached to this PR. Enable the module, assign it to e.g. the position-3 of the protostar template. Don't forget to set the module to show on all pages.

Depending on the settings (key, address) you've made, you will see a debug message with information about the last request. Now, there aren't really that much visible errors to indicate something not working correctly. The debug message in the module only show that the request works as before.

You can provoke the exception mentioned above by applying the patch and changing the https back to http on line 785 in /libraries/joomla/google/embed/maps.php. Given you've provided an api key you would see a message similar to this one:
![api_key_exception](https://cloud.githubusercontent.com/assets/1304616/15217884/ce56c7ce-185d-11e6-8d63-6501c8fb6474.jpg)
#### Attachment

[mod_geocodepr.zip](https://github.com/joomla/joomla-cms/files/261518/mod_geocodepr.zip)

matrikular committed:
Change api scheme to https

While the main change to this file is about setting the scheme to https, I've also refactored the uri handling and added the option to apply an api key as well.

matrikular committed:
Remove urlencode from address since it will be handled automatically in toString()

matrikular committed:
Create a new JUri instance instead of using $this->uri to prevent errors with e.g. isSercure()

matrikular committed:
Remove the sensor parameter since it doesn't seem to be need any longer

Remove the sensor parameter since it doesn't seem to be need any longer. Thanks for pointing it out @robert.

brianteeman commented:
@mbabker is this a PR that should be made to the framework as well / instead ?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10439">issues.joomla.org/joomla-cms/10439</a>.</sub>

matrikular committed:
Satisfy Travis, maybe?!

matrikular committed:
Add the sensor parameter back to the api url since the tests seem to rely on it

Add the sensor parameter back to the api url since the tests seem to rely on it. If that really is the case, we should implement be a better way of "parsing" the url and / or looking for an array index in the tests.

matrikular committed:
Fix url parsing to properly handle query params

matrikular committed:
Remove the sensor parameter since it doesn't seem to be needed any more

This time the test(s) shouldn't fail - fingers crossed.

kolvar commented:
Works as described. Fixes a problem I have had. Thanx for the patch.

bembelimen commented:
I have tested this item :white_check_mark: successfully on fd2305c99c99c4114ff23dc62c83dee8aed5a88c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10439">issues.joomla.org/joomla-cms/10439</a>.</sub>

brianteeman commented:
MERGER - I am setting this RTC but not sure if should be merged here or in the upstream package?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10439">issues.joomla.org/joomla-cms/10439</a>.</sub>

mbabker commented:
Yes it needs to be merged here because the CMS isn't using any of the HTTP adapter code (which includes all of the API packages) from the Framework.

brianteeman commented:
Thanks for confirming @mbabker

andrepereiradasilva commented:
IMHO it would be a good practice to make this change also in the joomla framework repository
https://github.com/joomla-framework/google-api/blob/master/src/Embed/Maps.php#L685

mbabker commented:
Should be fixed there, yes.  But since it's not a Composer pulled library here, it should be merged here and dealt with on the FW separately.

andrepereiradasilva commented:
@mbabker yes of course. This is RTC and can be merged.
The framework PR is another thing.

@matrikular can you do a PR also in the framework repository.
https://github.com/joomla-framework/google-api/blob/master/src/Embed/Maps.php#L685

matrikular commented:
Not sure if it has got the correct format, but - of course - here you go: [#10](https://github.com/joomla-framework/google-api/pull/10).

zero-24 commented:
@joomla-cms-bot. This here is one more to fix :P
